cluster-autoscaler/simulator/delta_cluster_snapshot.go

```diff
@@ -0,0 +1,523 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package simulator
+
+import (
+	"fmt"
+
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	schedulerlisters "k8s.io/kubernetes/pkg/scheduler/listers"
+	schedulernodeinfo "k8s.io/kubernetes/pkg/scheduler/nodeinfo"
+)
+
+// DeltaClusterSnapshot is implementation of ClusterSnapshot optimized for typical Cluster Autoscaler usage - (fork, add stuff, revert) repeated many times per loop.
+//
+// Complexity of some notable operations:
+//	fork - O(1)
+//	revert - O(1)
+//	commit - O(n)
+//	list all pods (no filtering) - O(n), cached
+//	list all pods (with filtering) - O(n)
+//	list node infos - O(n), cached
+//	get all nodes - O(n)
+//
+// Watch out for:
+//	node deletions, pod additions & deletions - invalidates cache of current snapshot
+//		(when forked affects delta, but not base.)
+//	pod affinity - causes scheduler framework to list pods with non-empty selector,
+//		so basic caching doesn't help.
+//
+type DeltaClusterSnapshot struct {
+	data *internalDeltaSnapshotData
+}
+
+type internalDeltaSnapshotDataNodeLister internalDeltaSnapshotData
+type internalDeltaSnapshotDataPodLister internalDeltaSnapshotData
+
+type internalDeltaSnapshotData struct {
+	baseData *internalDeltaSnapshotData
+
+	nodeInfoMap      map[string]*schedulernodeinfo.NodeInfo
+	deletedNodeInfos map[string]bool
+
+	nodeInfoList         []*schedulernodeinfo.NodeInfo
+	completeNodeInfoList []*schedulernodeinfo.NodeInfo
+	podList              []*apiv1.Pod
+	rawPods              [][]*apiv1.Pod
+	rawPodsTotal         int
+	havePodsWithAffinity []*schedulernodeinfo.NodeInfo
+}
+
+func (data *internalDeltaSnapshotData) GetNodeInfo(name string) (*schedulernodeinfo.NodeInfo, error) {
+	if data == nil {
+		return nil, fmt.Errorf("node not found")
+	}
+	if nodeInfo, found := data.nodeInfoMap[name]; found {
+		return nodeInfo, nil
+	}
+	if data.deletedNodeInfos[name] || data.baseData == nil {
+		return nil, fmt.Errorf("node not found")
+	}
+	return data.baseData.GetNodeInfo(name)
+}
+
+func (data *internalDeltaSnapshotData) GetNodeInfoList() []*schedulernodeinfo.NodeInfo {
+	if data == nil {
+		return nil
+	}
+	if data.completeNodeInfoList == nil {
+		data.buildNodeInfoList()
```

Comment:
it seems like `buildNodeInfoList` is always called in if checking if it not already built.
Maybe rename the function to `buildCompleteNodeInfoListIfNeeded` and move the check towards it.

Reply:
I really don't like the methods that doStuffIfTheStarsAreRight, but you're absolutely right this is such method.

Refactored to return the list instead. It also made testing easier :)